File: boj/dp/1463.cpp
#include <iostream>
#include <algorithm>

#define endl '\n'
#define fastio cin.sync_with_stdio(false); cin.tie(nullptr)

using namespace std;

const int MAX = 1000001;
int d[MAX];

int main(){
    fastio;

    int n;
    cin >> n;
    for(int i=2;i<=n;i++){
	d[i] = d[i-1] + 1;
	if(i % 2 == 0) d[i] = min(d[i], d[i/2] + 1);
	if(i % 3 == 0) d[i] = min(d[i], d[i/3] + 1);
    }
    cout << d[n] << endl;

    return 0;
}

File: boj/dp/1463.py
n = int(input())

dp = []
dp.append(0)
dp.append(0)

for i in range(2,n+1):
    dp.append(dp[i-1] + 1)
    if i % 2 == 0:
        dp[i] = min(dp[i], dp[i // 2] + 1)
    if i % 3 == 0:
        dp[i] = min(dp[i], dp[i // 3] + 1)

print(dp[n])
